README.md
# ratelimiter

[![NPM version](https://img.shields.io/npm/v/@zcong/ratelimiter.svg?style=flat)](https://npmjs.com/package/@zcong/ratelimiter) [![NPM downloads](https://img.shields.io/npm/dm/@zcong/ratelimiter.svg?style=flat)](https://npmjs.com/package/@zcong/ratelimiter) [![JS Test](https://github.com/zcong1993/ratelimiter/actions/workflows/js-test.yml/badge.svg)](https://github.com/zcong1993/ratelimiter/actions/workflows/js-test.yml) [![codecov](https://codecov.io/gh/zcong1993/ratelimiter/branch/master/graph/badge.svg)](https://codecov.io/gh/zcong1993/ratelimiter)

> my cool project

## License

MIT &copy; zcong1993


CHANGELOG.md
## [0.2.2](https://github.com/zcong1993/ratelimiter/compare/v0.2.1...v0.2.2) (2021-09-23)


### Bug Fixes

* lock rolling-policy version ([b3d2b35](https://github.com/zcong1993/ratelimiter/commit/b3d2b352a21516e393782a129df2258a979dea73))



## [0.2.1](https://github.com/zcong1993/ratelimiter/compare/v0.2.0...v0.2.1) (2020-04-30)


### Bug Fixes

* use passed cpuproc ([b5460d7](https://github.com/zcong1993/ratelimiter/commit/b5460d710d83dbd3af74006e6be88070414a91eb))



# [0.2.0](https://github.com/zcong1993/ratelimiter/compare/v0.1.0...v0.2.0) (2020-04-30)


### Features

* **group:** support group ([d49c1bf](https://github.com/zcong1993/ratelimiter/commit/d49c1bf601b543428ca4e6ea5b9d19f5a6b4db70))



# [0.1.0](https://github.com/zcong1993/ratelimiter/compare/549c2bb31d3c9d76f360a82d138600a886231892...v0.1.0) (2020-01-16)


### Features

* add cpu stat ([549c2bb](https://github.com/zcong1993/ratelimiter/commit/549c2bb31d3c9d76f360a82d138600a886231892))





__test__/bbr.test.ts
import { sleepMs, RollingCounter } from '@zcong/rolling-policy'
import { Bbr } from '../src/bbr'

const createBbr = () => {
  const bbr = new Bbr({
    cpuThreshold: 80,
    window: 1000,
    winBucket: 10,
  })

  return [bbr as any, () => bbr.destroy()]
}

const createBbrWithCpu = (cpu: any): any => {
  const bbr: any = new Bbr({
    cpuThreshold: 80,
    window: 1000,
    winBucket: 10,
  })
  bbr.destroy()
  bbr.cpu = cpu
  return bbr
}

const mockCpu = () => {
  class MockCpuproc {
    cpu: number = 0
    stat() {
      return this.cpu
    }
  }

  return new MockCpuproc()
}

const afterN = async (fn: Function, n: number) => {
  await sleepMs(n)
  return fn()
}

it('maxPass should works well', async () => {
  const [publicB, destroy] = createBbr()
  const d = 100
  for (let i = 0; i <= 10; i++) {
    publicB.passState.add(i * 100)
    await sleepMs(d)
  }

  expect(publicB.maxPass()).toBe(1000)

  destroy()
})

it('minRt should works well', async () => {
  const [publicB, destroy] = createBbr()
  const d = 100
  publicB.rtStat = new RollingCounter({ size: 10, bucketDuration: d })
  for (let i = 0; i < 10; i++) {
    for (let j = i * 10 + 1; j <= i * 10 + 10; j++) {
      publicB.rtStat.add(j)
    }
    if (i !== 9) {
      await sleepMs(d)
    }
  }

  expect(publicB.minRt()).toBe(6)

  destroy()
})

it('maxQps should works well', async () => {
  const [publicB, destroy] = createBbr()
  const d = 100
  for (let i = 0; i < 10; i++) {
    publicB.passState.add((i + 2) * 100)
    for (let j = i * 10 + 1; j <= i * 10 + 10; j++) {
      publicB.rtStat.add(j)
    }
    if (i !== 9) {
      await sleepMs(d)
    }
  }
  expect(publicB.maxFlight()).toBe(60)

  destroy()
})

it('should drop', async () => {
  const cpu = mockCpu()
  const publicB = createBbrWithCpu(cpu)
  const d = 100

  for (let i = 0; i < 10; i++) {
    publicB.passState.add((i + 1) * 100)
    for (let j = i * 10 + 1; j <= i * 10 + 10; j++) {
      publicB.rtStat.add(j)
    }
    if (i !== 9) {
      await sleepMs(d)
    }
  }

  // cpu >=  80, inflight < maxQps
  cpu.cpu = 80
  publicB.inFlight = 50
  expect(publicB.shouldDrop()).toBeFalsy()

  // cpu >=  80, inflight > maxQps
  cpu.cpu = 80
  publicB.inFlight = 80
  expect(publicB.shouldDrop()).toBeTruthy()

  // cpu < 80, inflight > maxQps, cold duration
  cpu.cpu = 70
  publicB.inFlight = 80
  expect(publicB.shouldDrop()).toBeTruthy()

  // cpu < 80, inflight > maxQps
  await sleepMs(2000)
  cpu.cpu = 70
  publicB.inFlight = 80
  expect(publicB.shouldDrop()).toBeFalsy()
})

it('bbr should works well', async () => {
  const bbr = new Bbr({
    window: 5000,
    winBucket: 50,
    cpuThreshold: 10,
  })

  let drop = 0

  const createFn = (n: number) => {
    return async () => {
      for (let i = 0; i < n; i++) {
        const rd = Math.floor(Math.random() * 100)
        try {
          const cb = bbr.allow()
          await sleepMs(rd)
          cb({ success: true })
        } catch (err) {
          drop++
        }
      }
    }
  }

  const fns = () =>
    Array(100)
      .fill(null)
      .map(() => createFn(300)())

  await Promise.all(fns())
  console.log(`drop: ${drop}`)
  bbr.destroy()
}, 20000)

it('test default config', async () => {
  const bbr = new Bbr()
  const cb = bbr.allow()
  cb({ success: true })
  bbr.destroy()
})

it('test stat', async () => {
  const bbr = new Bbr()
  const cb = bbr.allow()
  cb({ success: true })
  expect(bbr.stat).toBeDefined()
  bbr.destroy()
})


__test__/group.test.ts
import { Group } from '../src/group'

it('group should works well', () => {
  const mockRes = 'test'
  const newFunc = jest.fn<string, any>().mockReturnValue(mockRes)

  const g = new Group(newFunc)
  expect(g.get('key1')).toBe(mockRes)
  expect(newFunc).toBeCalledTimes(1)
  expect(g.get('key1')).toBe(mockRes)
  expect(newFunc).toBeCalledTimes(1)

  expect(g.get('key2')).toBe(mockRes)
  expect(newFunc).toBeCalledTimes(2)

  g.reset()
  expect(g.get('key1')).toBe(mockRes)
  expect(newFunc).toBeCalledTimes(3)
})


__test__/index.test.ts
import { Bbr } from '../src'

it('should work well', () => {
  expect(typeof Bbr).toBe('function')
})


src/cpu.ts
import * as os from 'os'

const getCpus = () => os.cpus()

const getAllBusy = (t: os.CpuInfo): [number, number] => {
  const idle: number = t.times.idle
  const busy: number = t.times.irq + t.times.nice + t.times.sys + t.times.user

  return [busy + idle, busy]
}

const calculateBusy = (t1: os.CpuInfo, t2: os.CpuInfo): number => {
  const [t1All, t1Busy] = getAllBusy(t1)
  const [t2All, t2Busy] = getAllBusy(t2)

  if (t2Busy <= t1Busy) {
    /* istanbul ignore next */
    return 0
  }

  if (t2All <= t1All) {
    /* istanbul ignore next */
    return 100
  }

  return Math.min(100, Math.max(0, ((t2Busy - t1Busy) / (t2All - t1All)) * 100))
}

const calculateAllBusy = (t1: os.CpuInfo[], t2: os.CpuInfo[]): number[] => {
  if (t1.length !== t2.length) {
    /* istanbul ignore next */
    throw new Error(`received two CPU counts: ${t1.length} != ${t2.length}`)
  }

  const res: number[] = []

  for (let i = 0; i < t1.length; i += 1) {
    res[i] = calculateBusy(t1[i], t2[i])
  }

  return res
}

const mergeAll = (t: os.CpuInfo[]): [os.CpuInfo] => {
  const res: os.CpuInfo = {
    model: '',
    speed: 0,
    times: {
      idle: 0,
      user: 0,
      sys: 0,
      irq: 0,
      nice: 0,
    },
  }

  for (const tt of t) {
    res.times.idle += tt.times.idle
    res.times.user += tt.times.user
    res.times.sys += tt.times.sys
    res.times.irq += tt.times.irq
    res.times.nice += tt.times.nice
  }

  return [res]
}

export class CpuStat {
  private isFirstTime: boolean = true
  private lastCpuInfo: os.CpuInfo[]
  private cpuPercent: number = 0
  private timer: NodeJS.Timeout

  constructor(private readonly sampleInterval: number = 500) {
    this.timer = setInterval(() => this.ticker(), this.sampleInterval)
  }

  ticker() {
    if (this.isFirstTime) {
      this.lastCpuInfo = getCpus()
      this.isFirstTime = false
      return
    }

    const nowInfo = getCpus()
    const res = calculateAllBusy(mergeAll(this.lastCpuInfo), mergeAll(nowInfo))
    this.cpuPercent = res[0]
    this.lastCpuInfo = nowInfo
  }

  get stat() {
    return this.cpuPercent
  }

  destroy() {
    if (this.timer) {
      clearInterval(this.timer)
    }
  }
}

export class Cpuproc {
  private lastCpu: number = 0
  private cpu: number = 0
  private timer: NodeJS.Timeout
  constructor(
    private readonly cpuStat: CpuStat,
    private readonly sampleInterval: number = 500,
    private readonly decay: number = 0.95
  ) {
    this.timer = setInterval(() => this.ticker(), this.sampleInterval)
  }

  ticker() {
    const curCpu = this.cpuStat.stat
    this.cpu = this.lastCpu * this.decay + curCpu * (1 - this.decay)
    this.lastCpu = curCpu
  }

  stat() {
    return this.cpu
  }

  destroy() {
    this.cpuStat.destroy()
    if (this.timer) {
      clearInterval(this.timer)
    }
  }
}


src/rollingCounter.ts
import * as debug from 'debug'

const debugLogger = debug('rollingCounter')

interface C {
  val: number
  time: number
  date?: Date
}

interface RollingCounterOpts {
  windowNano?: number
  gcInterval?: number
}

const ONE_SECOND_NANO = 1e9

const nano = (): number => {
  const hr = process.hrtime()
  return hr[0] * 1000000000 + hr[1]
}

const defaultOpts: RollingCounterOpts = {
  windowNano: ONE_SECOND_NANO,
  gcInterval: 5000
}

export class RollingCounter {
  private store: C[] = []
  private opt: RollingCounterOpts
  private gcTimer: NodeJS.Timeout

  constructor(option: RollingCounterOpts = {}) {
    this.opt = {
      ...defaultOpts,
      ...option
    }
    this.gcTimer = setInterval(() => this.gc(), this.opt.gcInterval)
  }

  add(val: number, time: number = nano()) {
    this.store.push({
      val,
      time
    })
  }

  reduceTotal(start?: number): number {
    const s = start || this.windowStartNano()
    this.gc(s)
    return this.store.reduce((prev, acc) => acc.val + prev, 0)
  }

  gc(start?: number) {
    const s = start || this.windowStartNano()
    const i = this.store.findIndex(c => c.time >= s)
    if (i < 0) {
      this.store = []
    } else {
      this.store = this.store.slice(i)
    }
    debugLogger(`gc, i: ${i}, store size: ${this.store.length}`)
  }

  destroy() {
    clearInterval(this.gcTimer)
  }

  stats() {
    return {
      size: this.store.length
    }
  }

  private windowStartNano() {
    return nano() - this.opt.windowNano
  }
}


src/bbr.ts
import { RollingCounter, hrtime } from '@zcong/rolling-policy'
import { Cpuproc, CpuStat } from './cpu'

export class ErrorLimitExceed extends Error {}

export interface DoneInfo {
  success: boolean
}

export type CallbackFunc = (doneInfo: DoneInfo) => void

export interface Config {
  window?: number // ms
  winBucket?: number
  cpuThreshold?: number
  cupProc?: Cpuproc
}

export const defaultConfig: Config = {
  window: 5000,
  winBucket: 50,
  cpuThreshold: 80,
}

const ONE_SECOND_NANO = 1e9
const INIT_HRTIME = hrtime()

export const sinceHrtime = (since: number): number => {
  return hrtime() - since
}

export const sum = (nums: number[]): number =>
  nums.reduce((prev: number, cur: number) => prev + cur, 0)

export class Bbr {
  private readonly config: Config
  private readonly cpu: Cpuproc
  private passState: RollingCounter
  private rtStat: RollingCounter
  private inFlight: number = 0
  private winBucketPerSec: number
  private prevDrop: number = 0
  private prevDropHit: boolean = false
  private rawMaxPass: number = 0
  private rawMinRt: number = 0

  constructor(cfg?: Config) {
    if (!cfg) {
      this.config = defaultConfig
    } else {
      this.config = cfg
    }

    if (!this.config.cupProc) {
      this.cpu = new Cpuproc(new CpuStat())
    } else {
      this.cpu = this.config.cupProc
    }

    const size = this.config.winBucket
    const bucketDuration = this.config.window / size

    this.passState = new RollingCounter({ size, bucketDuration })
    this.rtStat = new RollingCounter({ size, bucketDuration })
    this.winBucketPerSec = 1000 / bucketDuration
  }

  allow(): CallbackFunc {
    const drop = this.shouldDrop()
    if (drop) {
      throw new ErrorLimitExceed()
    }

    this.inFlight += 1
    const startTime = sinceHrtime(INIT_HRTIME)
    return (doneInfo: DoneInfo) => {
      const rt = (sinceHrtime(INIT_HRTIME) - startTime) / 1e6
      this.rtStat.add(rt)
      this.inFlight -= 1
      if (doneInfo.success) {
        this.passState.add(1)
      }
    }
  }

  destroy() {
    this.cpu.destroy()
  }

  get stat() {
    return {
      cpu: this.cpu.stat(),
      inFlight: this.inFlight,
      minRt: this.minRt(),
      maxPass: this.maxPass(),
      maxInFlight: this.maxFlight(),
    }
  }

  private shouldDrop(): boolean {
    if (this.cpu.stat() < this.config.cpuThreshold) {
      if (this.prevDrop === 0) {
        return false
      }

      if (sinceHrtime(INIT_HRTIME) - this.prevDrop <= ONE_SECOND_NANO) {
        if (!this.prevDropHit) {
          this.prevDropHit = true
        }
        return this.inFlight > 1 && this.inFlight > this.maxFlight()
      }

      this.prevDrop = 0
      return false
    }

    const drop = this.inFlight > 1 && this.inFlight > this.maxFlight()
    if (drop) {
      if (this.prevDrop !== 0) {
        return drop
      }

      this.prevDrop = sinceHrtime(INIT_HRTIME)
    }

    return drop
  }

  private maxFlight(): number {
    return Math.floor(
      (this.maxPass() * this.minRt() * this.winBucketPerSec) / 1000 + 0.5
    )
  }

  private maxPass(): number {
    if (this.rawMaxPass > 0 && this.passState.timespan() < 1) {
      return this.rawMaxPass
    }

    let rawMaxPass = 1
    const currentPoints = this.passState.currentPoints()
    if (currentPoints.length > this.config.winBucket - 1) {
      currentPoints.pop()
    }
    for (const points of currentPoints) {
      rawMaxPass = Math.max(rawMaxPass, sum(points))
    }

    this.rawMaxPass = rawMaxPass
    return rawMaxPass
  }

  private minRt(): number {
    if (this.rawMinRt > 0 && this.rtStat.timespan() < 1) {
      return this.rawMinRt
    }

    let rawMinRt = Number.MAX_SAFE_INTEGER
    const currentBuckets = this.rtStat.currentBuckets()
    if (currentBuckets.length > this.config.winBucket - 1) {
      currentBuckets.pop()
    }
    for (const bucket of currentBuckets) {
      if (bucket.getCount() === 0) {
        continue
      }
      rawMinRt = Math.min(rawMinRt, sum(bucket.getPoints()) / bucket.getCount())
    }

    if (rawMinRt <= 0) {
      /* istanbul ignore next */
      rawMinRt = 1
    }

    this.rawMinRt = Math.ceil(rawMinRt)

    return this.rawMinRt
  }
}
